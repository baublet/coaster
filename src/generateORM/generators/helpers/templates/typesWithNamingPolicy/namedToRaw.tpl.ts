export function $$namedToRawFunctionName<
  T extends $$namedEntityName | Partial<$$namedEntityName>
>(subject: T): $$namedToRawReturnTypeSignature {
  const rawSubject: Record<string, any> = {};
  $$namedToRawPropertyAssignments;
  return rawSubject as $$namedToRawReturnTypeSignature;
}
