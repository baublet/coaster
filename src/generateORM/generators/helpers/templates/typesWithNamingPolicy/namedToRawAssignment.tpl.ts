  if (subject["$$namedColumnName"] !== undefined){
    rawSubject["$$rawColumnName"] = subject["$$namedColumnName"];
  }
