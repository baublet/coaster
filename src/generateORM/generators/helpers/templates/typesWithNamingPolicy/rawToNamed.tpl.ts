export function $$rawToNamedFunctionName<
  T extends $$rawEntityName | Partial<$$rawEntityName>
>(subject: T): $$rawToNamedReturnTypeSignature {
  const namedSubject: Record<string, any> = {};
$$rawToNamedPropertyAssignments
  return namedSubject as $$rawToNamedReturnTypeSignature;
}
