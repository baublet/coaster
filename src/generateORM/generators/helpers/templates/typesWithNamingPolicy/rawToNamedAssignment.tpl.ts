  if (subject["$$rawColumnName"] !== undefined) {
    namedSubject["$$namedColumnName"] = subject["$$rawColumnName"];
  }
