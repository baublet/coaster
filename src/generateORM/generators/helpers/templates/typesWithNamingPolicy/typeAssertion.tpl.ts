export function $$assertionFunctionName(
  subject: any
): asserts subject is $$prefixedEntityName {
  if (typeof subject === "object") {
    if (objectHasProperties(subject, $$columnNamesAsJsonString)) {
      return;
    }
  }
  throw new Error(
    "Invariance violation! Expected subject to be a $$prefixedEntityName, but it was instead: " +
      JSON.stringify(subject)
  );
}
