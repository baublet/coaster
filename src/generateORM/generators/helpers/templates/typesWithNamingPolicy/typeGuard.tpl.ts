export function $$typeGuardFunctionName(
  subject: any
): subject is $$prefixedEntityName {
  if (typeof subject === "object") {
    if (objectHasProperties(subject, $$columnNamesAsJsonString)) {
      return true;
    }
  }
  return false;
}
