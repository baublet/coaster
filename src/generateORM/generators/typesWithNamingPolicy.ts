import { camelCase } from "change-case";

import { orDefault, ternary } from "../../helpers";
import { MetaData, GetTypeName, GeneratorResult } from ".";
import { RawSchema } from "../drivers";
import { getSchemaAndTablePath, getName, getTypeName } from "./helpers";
import { generateNames } from "../../generateNames";

const defaultEntityNamingPolicy = (str: string) =>
  generateNames(str).singularPascal;
const defaultPropertyNamingPolicy = (str: string) =>
  generateNames(str).rawCamel;

/**
 * JS code often has different name conventions than database names (e.g., your
 * database might use snake_case, but your JS code might use camelCase). This
 * generator creates types, assertions, guards, and transformers to transform
 * raw DB types to any custom type structure and back again.
 */
export const typesWithNamingPolicy = (
  schema: RawSchema,
  metaData: MetaData,
  options: {
    enumPrefix?: string;
    typesOrInterfaces: "types" | "interfaces";
    prefixSchemaName?: boolean;
    prefix?: string;
    getTypeName?: GetTypeName;
    getEntityName: (tableName: string, schemaName: string) => string;
    getPropertyName: (
      propertyName: string,
      tableName: string,
      schemaName: string
    ) => string;
  } = {
    enumPrefix: "Enum",
    typesOrInterfaces: "interfaces",
    getEntityName: defaultEntityNamingPolicy,
    getPropertyName: defaultPropertyNamingPolicy,
  }
): GeneratorResult => {
  let code = "";
  let testCode = "";
  const prefix = orDefault([options.prefix], "");

  // Enums
  const schemaName = ternary(
    options.prefixSchemaName,
    generateNames(
      getName(undefined, undefined, schema.name, options.prefixSchemaName)
    ).singularPascal,
    ""
  );
  const enumPrefix = orDefault([options.enumPrefix], "Enum");
  for (const { name, values } of schema.enums) {
    const enumNames = generateNames(name);
    const enumTypeName = `${prefix}${enumNames.singularPascal}${schemaName}${enumPrefix}`;
    code += `export type ${enumTypeName} = `;
    code += '"' + values.join('" | "') + '"';
    code += `;\n`;
    metaData.namedDatabaseEnumNames.set(`${schema.name}.${name}`, enumTypeName);
  }

  // Entities
  for (const table of schema.tables) {
    const entityName = options.getEntityName(table.name, schema.name);
    const prefixedEntityName = prefix + entityName;

    const schemaAndTablePath = getSchemaAndTablePath(schema.name, table.name);
    metaData.entityTableNames.set(prefixedEntityName, schemaAndTablePath);
    metaData.tableEntityNames.set(schemaAndTablePath, prefixedEntityName);

    const requiredColumnNames: string[] = [];

    let columnsCode = "";
    for (const column of table.columns) {
      const columnName = options.getPropertyName(
        column.name,
        table.name,
        schema.name
      );

      columnsCode += metaData.templateManager.render({
        template: "typesWithNamingPolicy/entityProperty",
        variables: {
          columnName,
          columnTypeName: getTypeName({
            column,
            metaData,
            schema,
            table,
            getTypeName: options.getTypeName,
            rawOrNamed: "named",
          }),
          comment: column.comment ? `\n/** ${column.comment} */` : "",
          propertyNameTerminator: column.nullable ? "?: " : ": ",
        },
      });

      if (!column.nullable) {
        requiredColumnNames.push(columnName);
      }
    }

    code += metaData.templateManager.render({
      template: "typesWithNamingPolicy/entity",
      variables: {
        columns: columnsCode,
        comment: table.comment ? `/** ${table.comment} */` : "",
        entityName: prefixedEntityName,
        interfaceOrType:
          options.typesOrInterfaces === "interfaces" ? "interface" : "type",
        typeEqualsSign: options.typesOrInterfaces === "interfaces" ? "" : "= ",
      },
    });

    const columnNamesAsJsonString = JSON.stringify(requiredColumnNames);

    // Type assertions
    const assertionFunctionName = `assertIs${prefix}${entityName}Like`;
    code += metaData.templateManager.render({
      template: "typesWithNamingPolicy/typeAssertion",
      variables: {
        assertionFunctionName,
        prefixedEntityName,
        columnNamesAsJsonString,
      },
    });

    metaData.typeAssertionFunctionNames.set(
      schemaAndTablePath,
      assertionFunctionName
    );

    if (metaData.generateTestCode) {
      testCode += metaData.templateManager.render({
        template: "typesWithNamingPolicy/typeAssertion.test",
        variables: {
          assertionFunctionName,
          codeOutputFullPath: metaData.codeOutputFullPath,
          entityLikeAsJsonString: `{${requiredColumnNames
            .map((col) => `${col}: ""`)
            .join(",")}}`,
          prefixedEntityName,
        },
      });
    }

    // Type guards
    const typeGuardFunctionName = `is${prefixedEntityName}Like`;

    code += metaData.templateManager.render({
      template: "typesWithNamingPolicy/typeGuard",
      variables: {
        columnNamesAsJsonString,
        prefixedEntityName,
        typeGuardFunctionName,
      },
    });

    metaData.typeGuardFunctionNames.set(
      schemaAndTablePath,
      typeGuardFunctionName
    );

    if (metaData.generateTestCode) {
      testCode += metaData.templateManager.render({
        template: "typesWithNamingPolicy/typeGuard.test",
        variables: {
          codeOutputFullPath: metaData.codeOutputFullPath,
          entityLikeAsJsonString: `{${requiredColumnNames
            .map((col) => `${col}: ""`)
            .join(",")}}`,
          typeGuardFunctionName,
          prefixedEntityName,
        },
      });
    }

    // Transformers
    const rawEntityName = metaData.tableRawEntityNames.get(schemaAndTablePath);
    const namedEntityName = prefixedEntityName;

    // Raw to named
    const rawToNamedFunctionName = camelCase(
      `${rawEntityName}To${namedEntityName}`
    );

    const rawToNamedPropertyAssignments = table.columns
      .map((column) => {
        const namedColumnName = options.getPropertyName(
          column.name,
          table.name,
          schema.name
        );

        metaData.namedEntityColumnNames.set(
          `${schemaAndTablePath}.${column.name}`,
          namedColumnName
        );

        return metaData.templateManager.render({
          template: "typesWithNamingPolicy/rawToNamedAssignment",
          variables: {
            namedColumnName: namedColumnName,
            rawColumnName: column.name,
          },
        });
      })
      .join("");

    code += metaData.templateManager.render({
      template: "typesWithNamingPolicy/rawToNamed",
      variables: {
        rawEntityName,
        rawToNamedFunctionName,
        rawToNamedPropertyAssignments,
        rawToNamedReturnTypeSignature: `T extends ${rawEntityName} ? ${namedEntityName} : Partial<${namedEntityName}>`,
      },
    });

    metaData.transformerFunctionNames[rawEntityName] =
      metaData.transformerFunctionNames[rawEntityName] || {};
    metaData.transformerFunctionNames[rawEntityName][namedEntityName] =
      rawToNamedFunctionName;

    if (metaData.generateTestCode) {
      testCode += metaData.templateManager.render({
        template: "typesWithNamingPolicy/rawToNamed.test",
        variables: {
          codeOutputFullPath: metaData.codeOutputFullPath,
          namedEntityName,
          namedFullTestSubject: `{${Object.values(table.columns)
            .map(
              (col) =>
                `${metaData.namedEntityColumnNames.get(
                  `${schemaAndTablePath}.${col.name}`
                )}: "" as any`
            )
            .join(",")}}`,
          rawEntityName,
          rawFullTestSubject: `{${Object.values(table.columns)
            .map((col) => `${col.name}: "" as any`)
            .join(",")}}`,
          rawToNamedFunctionName,
        },
      });
    }

    // Named to raw
    const namedToRawFunctionName = camelCase(
      `${namedEntityName}To${rawEntityName}`
    );

    const namedToRawReturnTypeSignature = `T extends ${entityName} ? ${rawEntityName} : Partial<${rawEntityName}>`;
    code += `export function ${namedToRawFunctionName}`;
    code += `<T extends ${namedEntityName} | Partial<${namedEntityName}>>(subject: T): ${namedToRawReturnTypeSignature} {\n`;
    code += `  const rawSubject: Record<string, any> = {};\n`;
    for (const column of table.columns) {
      const columnName = options.getPropertyName(
        column.name,
        table.name,
        schema.name
      );
      code += `    if(subject["${columnName}"] !== undefined) rawSubject["${column.name}"] = subject["${columnName}"];\n`;
    }
    code += `  return rawSubject as ${namedToRawReturnTypeSignature};\n`;
    code += `}\n\n`;

    metaData.transformerFunctionNames[namedEntityName] =
      metaData.transformerFunctionNames[namedEntityName] || {};
    metaData.transformerFunctionNames[namedEntityName][rawEntityName] =
      namedToRawFunctionName;

    if (metaData.generateTestCode) {
      testCode += `\n\nimport { ${namedToRawFunctionName} } from "${
        metaData.codeOutputFullPath
      }";
  
describe("${namedToRawFunctionName}", () => {
  const ${namedEntityName}Full: ${namedEntityName} = {
    ${Object.values(table.columns)
      .map(
        (col) =>
          `${metaData.namedEntityColumnNames.get(
            `${schemaAndTablePath}.${col.name}`
          )}: "" as any`
      )
      .join(",")}
  };
  const ${rawEntityName}Full: ${rawEntityName} = {
    ${Object.values(table.columns)
      .map((col) => `${col.name}: "" as any`)
      .join(",")}
  };
  const ${rawEntityName}Partial: Partial<${rawEntityName}> = {};
  const ${namedEntityName}Partial: Partial<${namedEntityName}> = {};
  it("converts a full ${namedEntityName} to a full ${rawEntityName}", () => {
    expect(${namedToRawFunctionName}(${namedEntityName}Full)).toEqual(${rawEntityName}Full);
  });
  it("converts a partial ${namedEntityName} to a partial ${rawEntityName}", () => {
    expect(${namedToRawFunctionName}(${namedEntityName}Partial)).toEqual(${rawEntityName}Partial);
  });
});`;
    }

    // Named input type -- all nullable and non-nullable fields with defaults
    // are optional
    code += `export ${
      options.typesOrInterfaces === "interfaces" ? "interface" : "type"
    } ${prefixedEntityName}Input ${
      options.typesOrInterfaces === "interfaces" ? "" : "= "
    }{`;
    for (const column of table.columns) {
      const columnName = options.getPropertyName(
        column.name,
        table.name,
        schema.name
      );
      if (column.comment) {
        code += `\n/** ${column.comment} */`;
      }
      code += `\n${columnName}`;
      code +=
        column.nullable === true || column.hasDefault === true ? "?: " : ": ";

      code += getTypeName({
        column,
        metaData,
        schema,
        table,
        getTypeName: options.getTypeName,
        rawOrNamed: "named",
      });

      code += ";";
      if (!column.nullable) {
        requiredColumnNames.push(columnName);
      }
      metaData.namedEntityInputTypeNames.set(
        schemaAndTablePath,
        `${prefixedEntityName}Input`
      );
    }
    code += `\n}\n\n`;
  }

  return {
    code,
    testCode,
  };
};
