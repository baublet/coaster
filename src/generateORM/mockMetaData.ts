import { MetaData } from "./generators";
import { getTemplateManager } from "../generateORM/generators/helpers";

export function getMockMetaData(): MetaData {
  return {
    testConnectionVariable: "connection",
    codeOutputFullPath: "./generated.ts",
    generateTestCode: false,
    entityTableNames: new Map(),
    setHeader: jest.fn(),
    setTestHeader: jest.fn(),
    tableEntityNames: new Map(),
    rawBaseQueryFunctionNames: new Map(),
    transformerFunctionNames: {},
    typeAssertionFunctionNames: new Map(),
    typeGuardFunctionNames: new Map(),
    tableRawEntityNames: new Map(),
    namedEntityColumnNames: new Map(),
    namedEntityInputTypeNames: new Map(),
    namedDatabaseEnumNames: new Map(),
    rawDatabaseEnumNames: new Map(),
    namedCreateTestEntityFunctionNames: new Map(),
    rawCreateTestEntityFunctionNames: new Map(),
    rawEnumValues: new Map(),
    templateManager: getTemplateManager(),
    namedMockEntityFunctionNames: new Map(),
    rawMockEntityFunctionNames: new Map(),
  };
}
