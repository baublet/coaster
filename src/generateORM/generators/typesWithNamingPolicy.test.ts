import { pascalCase, camelCase } from "change-case";

import { typesWithNamingPolicy } from "./typesWithNamingPolicy";
import { getMockRawSchema } from "../mockRawSchema";
import { getMockMetaData } from "../mockMetaData";

const entityNamingPolicy = (str: string) => pascalCase(str);
const propertyNamingPolicy = (str: string) => camelCase(str);

const mockMetaData = getMockMetaData();
mockMetaData.tableRawEntityNames.set("public.user", "RawUser");
mockMetaData.tableRawEntityNames.set("public.user_account", "RawUserAccount");

describe("interfaces/types", () => {
  it("spits out types with the proper naming policy", () => {
    expect(
      typesWithNamingPolicy(getMockRawSchema(), mockMetaData, {
        typesOrInterfaces: "types",
        getEntityName: entityNamingPolicy,
        getPropertyName: propertyNamingPolicy,
      })
    ).toMatchSnapshot();
  });
});

describe("custom type names: `string` to `CHUMBAWUMBA`", () => {
  it("generates types with the proper naming policy", () => {
    expect(
      typesWithNamingPolicy(getMockRawSchema(), mockMetaData, {
        typesOrInterfaces: "types",
        getEntityName: entityNamingPolicy,
        getPropertyName: propertyNamingPolicy,
        getTypeName: (typeName) => {
          if (typeName === "string") {
            return "CHUMBAWUMBA";
          }
        },
      })
    ).toMatchSnapshot();
  });
});

it("generates test code", () => {
  mockMetaData.generateTestCode = true;
  expect(
    typesWithNamingPolicy(getMockRawSchema(), mockMetaData)
  ).toMatchSnapshot();
});
