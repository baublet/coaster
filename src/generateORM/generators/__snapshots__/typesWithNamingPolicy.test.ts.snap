// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom type names: \`string\` to \`CHUMBAWUMBA\` spits out types with the proper naming policy 1`] = `
"export type User = {
id: CHUMBAWUMBA;
/** Users don't have to provide their name */
name?: CHUMBAWUMBA;
data: JSON;
};

export function assertIsUserLike(subject: any): asserts subject is User {
  if(typeof subject === \\"object\\") {
    if(objectHasProperties(subject, [\\"id\\",\\"data\\"])) { return; }
  }
  throw new Error(\\"Invariance violation! Expected subject to be a User, but it was instead: \\" + JSON.stringify(subject));
}

export function isUserLike(subject: any): subject is User {
  if(typeof subject === \\"object\\") {
    if(objectHasProperties(subject, [\\"id\\",\\"data\\"])) { return true; }
  }
  return false;
}

export function rawUserToUser<T extends RawUser | Partial<RawUser>>(subject: T): T extends RawUser ? User : Partial<User> {
  const namedSubject: Record<string, any> = {};
    if(subject[\\"id\\"] !== undefined) namedSubject[\\"id\\"] = subject[\\"id\\"];
    if(subject[\\"name\\"] !== undefined) namedSubject[\\"name\\"] = subject[\\"name\\"];
    if(subject[\\"data\\"] !== undefined) namedSubject[\\"data\\"] = subject[\\"data\\"];
  return namedSubject as T extends RawUser ? User : Partial<User>;
}

export function userToRawUser<T extends User | Partial<User>>(subject: T): T extends User ? RawUser : Partial<RawUser> {
  const rawSubject: Record<string, any> = {};
    if(subject[\\"id\\"] !== undefined) rawSubject[\\"id\\"] = subject[\\"id\\"];
    if(subject[\\"name\\"] !== undefined) rawSubject[\\"name\\"] = subject[\\"name\\"];
    if(subject[\\"data\\"] !== undefined) rawSubject[\\"data\\"] = subject[\\"data\\"];
  return rawSubject as T extends User ? RawUser : Partial<RawUser>;
}

export type UserInput = {
id?: CHUMBAWUMBA;
/** Users don't have to provide their name */
name?: CHUMBAWUMBA;
data: JSON;
}


/** Singular users can have multiple accounts */
export type UserAccount = {
id: unknown;
/** Who sent us this users' data? */
source: CHUMBAWUMBA;
/** The raw data sent from the source provider */
sourceData: JSON;
};

export function assertIsUserAccountLike(subject: any): asserts subject is UserAccount {
  if(typeof subject === \\"object\\") {
    if(objectHasProperties(subject, [\\"id\\",\\"source\\",\\"sourceData\\"])) { return; }
  }
  throw new Error(\\"Invariance violation! Expected subject to be a UserAccount, but it was instead: \\" + JSON.stringify(subject));
}

export function isUserAccountLike(subject: any): subject is UserAccount {
  if(typeof subject === \\"object\\") {
    if(objectHasProperties(subject, [\\"id\\",\\"source\\",\\"sourceData\\"])) { return true; }
  }
  return false;
}

export function rawUserAccountToUserAccount<T extends RawUserAccount | Partial<RawUserAccount>>(subject: T): T extends RawUserAccount ? UserAccount : Partial<UserAccount> {
  const namedSubject: Record<string, any> = {};
    if(subject[\\"id\\"] !== undefined) namedSubject[\\"id\\"] = subject[\\"id\\"];
    if(subject[\\"source\\"] !== undefined) namedSubject[\\"source\\"] = subject[\\"source\\"];
    if(subject[\\"source_data\\"] !== undefined) namedSubject[\\"sourceData\\"] = subject[\\"source_data\\"];
  return namedSubject as T extends RawUserAccount ? UserAccount : Partial<UserAccount>;
}

export function userAccountToRawUserAccount<T extends UserAccount | Partial<UserAccount>>(subject: T): T extends UserAccount ? RawUserAccount : Partial<RawUserAccount> {
  const rawSubject: Record<string, any> = {};
    if(subject[\\"id\\"] !== undefined) rawSubject[\\"id\\"] = subject[\\"id\\"];
    if(subject[\\"source\\"] !== undefined) rawSubject[\\"source\\"] = subject[\\"source\\"];
    if(subject[\\"sourceData\\"] !== undefined) rawSubject[\\"source_data\\"] = subject[\\"sourceData\\"];
  return rawSubject as T extends UserAccount ? RawUserAccount : Partial<RawUserAccount>;
}

export type UserAccountInput = {
id?: unknown;
/** Who sent us this users' data? */
source: CHUMBAWUMBA;
/** The raw data sent from the source provider */
sourceData: JSON;
}

"
`;

exports[`interfaces/types spits out types with the proper naming policy 1`] = `
"export type User = {
id: string;
/** Users don't have to provide their name */
name?: string;
data: JSON;
};

export function assertIsUserLike(subject: any): asserts subject is User {
  if(typeof subject === \\"object\\") {
    if(objectHasProperties(subject, [\\"id\\",\\"data\\"])) { return; }
  }
  throw new Error(\\"Invariance violation! Expected subject to be a User, but it was instead: \\" + JSON.stringify(subject));
}

export function isUserLike(subject: any): subject is User {
  if(typeof subject === \\"object\\") {
    if(objectHasProperties(subject, [\\"id\\",\\"data\\"])) { return true; }
  }
  return false;
}

export function rawUserToUser<T extends RawUser | Partial<RawUser>>(subject: T): T extends RawUser ? User : Partial<User> {
  const namedSubject: Record<string, any> = {};
    if(subject[\\"id\\"] !== undefined) namedSubject[\\"id\\"] = subject[\\"id\\"];
    if(subject[\\"name\\"] !== undefined) namedSubject[\\"name\\"] = subject[\\"name\\"];
    if(subject[\\"data\\"] !== undefined) namedSubject[\\"data\\"] = subject[\\"data\\"];
  return namedSubject as T extends RawUser ? User : Partial<User>;
}

export function userToRawUser<T extends User | Partial<User>>(subject: T): T extends User ? RawUser : Partial<RawUser> {
  const rawSubject: Record<string, any> = {};
    if(subject[\\"id\\"] !== undefined) rawSubject[\\"id\\"] = subject[\\"id\\"];
    if(subject[\\"name\\"] !== undefined) rawSubject[\\"name\\"] = subject[\\"name\\"];
    if(subject[\\"data\\"] !== undefined) rawSubject[\\"data\\"] = subject[\\"data\\"];
  return rawSubject as T extends User ? RawUser : Partial<RawUser>;
}

export type UserInput = {
id?: string;
/** Users don't have to provide their name */
name?: string;
data: JSON;
}


/** Singular users can have multiple accounts */
export type UserAccount = {
id: unknown;
/** Who sent us this users' data? */
source: string;
/** The raw data sent from the source provider */
sourceData: JSON;
};

export function assertIsUserAccountLike(subject: any): asserts subject is UserAccount {
  if(typeof subject === \\"object\\") {
    if(objectHasProperties(subject, [\\"id\\",\\"source\\",\\"sourceData\\"])) { return; }
  }
  throw new Error(\\"Invariance violation! Expected subject to be a UserAccount, but it was instead: \\" + JSON.stringify(subject));
}

export function isUserAccountLike(subject: any): subject is UserAccount {
  if(typeof subject === \\"object\\") {
    if(objectHasProperties(subject, [\\"id\\",\\"source\\",\\"sourceData\\"])) { return true; }
  }
  return false;
}

export function rawUserAccountToUserAccount<T extends RawUserAccount | Partial<RawUserAccount>>(subject: T): T extends RawUserAccount ? UserAccount : Partial<UserAccount> {
  const namedSubject: Record<string, any> = {};
    if(subject[\\"id\\"] !== undefined) namedSubject[\\"id\\"] = subject[\\"id\\"];
    if(subject[\\"source\\"] !== undefined) namedSubject[\\"source\\"] = subject[\\"source\\"];
    if(subject[\\"source_data\\"] !== undefined) namedSubject[\\"sourceData\\"] = subject[\\"source_data\\"];
  return namedSubject as T extends RawUserAccount ? UserAccount : Partial<UserAccount>;
}

export function userAccountToRawUserAccount<T extends UserAccount | Partial<UserAccount>>(subject: T): T extends UserAccount ? RawUserAccount : Partial<RawUserAccount> {
  const rawSubject: Record<string, any> = {};
    if(subject[\\"id\\"] !== undefined) rawSubject[\\"id\\"] = subject[\\"id\\"];
    if(subject[\\"source\\"] !== undefined) rawSubject[\\"source\\"] = subject[\\"source\\"];
    if(subject[\\"sourceData\\"] !== undefined) rawSubject[\\"source_data\\"] = subject[\\"sourceData\\"];
  return rawSubject as T extends UserAccount ? RawUserAccount : Partial<RawUserAccount>;
}

export type UserAccountInput = {
id?: unknown;
/** Who sent us this users' data? */
source: string;
/** The raw data sent from the source provider */
sourceData: JSON;
}

"
`;
